skbio/io/format/gff3.py

```diff
@@ -0,0 +1,466 @@
+>>> im_return   # doctest: +SKIP
+3 interval features
+-------------------
+Interval(interval_metadata=<4601272528>, bounds=[(999, 9000)], fuzzy=\
+[(False, False)], metadata={'source': '.', 'type': 'gene', 'strand': '+', \
+'score': '.', 'phase': 0, 'ID': 'gene00001', 'Name': 'EDEN'})
+Interval(interval_metadata=<4601272528>, bounds=[(999, 1012)], fuzzy=\
+[(False, False)], metadata={'source': '.', 'type': 'TF_binding_site', \
+'strand': '+', 'score': '.', 'Parent': 'gene00001'})
+Interval(interval_metadata=<4601272528>, bounds=[(1049, 9000)], fuzzy=\
+[(False, False)], metadata={'source': '.', 'type': 'mRNA', 'strand': '+', \
+'score': '.', 'ID': 'mRNA00001', 'Parent': 'gene00001'})
+>>> im == im_return
+True
+
+References
+----------
+.. [1] https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md  # noqa
+
+'''
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import re
+from io import StringIO
+
+from skbio.sequence import DNA, Sequence
+from skbio.io import create_format, GFF3FormatError
+from skbio.metadata import IntervalMetadata
+from skbio.io.format._base import (
+    _line_generator, _too_many_blanks)
+from skbio.io.format._base import _get_nth_sequence as _get_nth_record
+from skbio.io.format._sequence_feature_vocabulary import (
+    _vocabulary_change, _vocabulary_skip)
+from skbio.io import write, read
+
+
+gff3 = create_format('gff3')
+
+
+@gff3.sniffer()
+def _gff3_sniffer(fh):
+    # check the 1st real line is a valid ID line
+    if _too_many_blanks(fh, 5):
+        return False, {}
+
+    try:
+        line = next(_line_generator(fh, skip_blanks=True, strip=False))
+    except StopIteration:
+        return False, {}
+
+    if re.match(r'##gff-version\s+3', line):
+        return True, {}
+    else:
+        return False, {}
+
+
+@gff3.reader(None)
+def _gff3_to_generator(fh, lengths):
+    '''Parse the GFF3 into the existing IntervalMetadata
+
+    Parameters
+    ----------
+    fh : file
+        file handler
+
+    Yields
+    ------
+    tuple
+        str of seq id, IntervalMetadata
+    '''
+    for (seq_id, lines), length in zip(_yield_record(fh), lengths):
+        yield seq_id, _parse_record(lines, length)
+
+
+@gff3.writer(None)
+def _generator_to_gff3(obj, fh, seq_ids, skip=True):
+    '''Write list of IntervalMetadata into file.
+
+    Parameters
+    ----------
+    obj : Iterable of IntervalMetadata
+    fh : file handler
+    seq_ids : Iterable of seq id (str)
+    '''
+    for obj_i, seq_id in zip(obj, seq_ids):
+        _serialize_interval_metadata(obj_i, seq_id, fh, skip)
+
+
+@gff3.reader(Sequence)
+def _gff3_to_sequence(fh, rec_num=1):
```

Comment:
So this `rec_num` would refer to the record number in the FASTA portion of the file? I think that makes sense, probably rename to `seq_num` to consistency (and it's more explicit). This writer could also accept a `seq_id` (that's mutually exclusive with `rec_num`) if you think that's useful.